packages/core/src/lib/ConnectionManager.ts

```diff
@@ -0,0 +1,336 @@
+import { Connection } from "@libp2p/interface-connection";
+import { PeerId } from "@libp2p/interface-peer-id";
+import { Peer } from "@libp2p/interface-peer-store";
+import { IRelay, Tags } from "@waku/interfaces";
+import debug from "debug";
+import { Libp2p } from "libp2p";
+
+import { createEncoder } from "../index.js";
+
+import { RelayPingContentTopic } from "./relay/constants.js";
+
+const log = debug("waku:connection-manager");
+
+const DEFAULT_PEER_DISCOVERY_CONNECTION_INTERVAL = 10 * 1000;
+const DEFAULT_MAX_BOOTSTRAP_PEERS_ALLOWED = 1;
+const DEFAULT_DIAL_ATTEMPTS_BEFORE_BOOTSTRAP_CONNECTION = 5;
+
+export interface Options {
+  relayKeepAlive: number;
+  pingKeepAlive: number;
+  /**
+   * Max number of bootstrap peers allowed to be connected to, initially
+   * This is used to increase intention of dialing non-bootstrap peers, found using other discovery mechanisms (like Peer Exchange)
+   * Is overridden by @link dialAttemptsBeforeBootstrapConnection
+   */
+  maxBootstrapPeersAllowed?: number;
+  /**
+   * Factor by which the peer discovery interval is increased after every attempt
+   * This is used to reduce the frequency of attempts to connect to nodes
+   */
+  peerDiscoveryIntervalFactor?: number;
+  /**
+   * Number of attempts before dialing all available bootstrap nodes
+   * This is only used when other discovery mechanisms haven't yeild peers that are dialable
+   * This increases relative centralisation of the network
+   */
+  dialAttemptsBeforeBootstrapConnection?: number;
+}
+
+export interface UpdatedStates {
+  allPeers: Peer[];
+  allDialablePeers: Peer[];
+  bootstrapPeers: Peer[];
+  nonBootstrapPeers: Peer[];
+  dialableBootstrapPeers: Peer[];
+  dialableNonBootstrapPeers: Peer[];
+  allConnections: Connection[];
+  bootstrapConnections: Connection[];
+  nonBootstrapConnections: Connection[];
+}
+
+/**
+ * ConnectionManager is a singleton class that manages different steps in a libp2p connection lifecycle
+ */
+export class ConnectionManager {
+  private static instance: ConnectionManager;
+  private libp2p: Libp2p;
+  private relay?: IRelay;
+  private pingKeepAliveTimers: {
+    [peer: string]: ReturnType<typeof setInterval>;
+  };
+  private relayKeepAliveTimers: {
+    [peer: string]: ReturnType<typeof setInterval>;
+  };
+  private dialAttempt: number;
+
+  public isConnectionServiceStarted = false;
+
+  public static create(libp2p: Libp2p, options: Options): ConnectionManager {
+    if (!ConnectionManager.instance) {
+      ConnectionManager.instance = new ConnectionManager(libp2p, options);
+    }
+    return ConnectionManager.instance;
+  }
+
+  private constructor(libp2p: Libp2p, options: Options, relay?: IRelay) {
+    this.libp2p = libp2p;
```

Comment:
Instead of passing the whole `libp2p` object, I'd recommend to pass the necessary components. The dependency injection makes it easier to handle breaking changes from libp2p interfaces.
Similar to what is done with the protocols (store, etc).